src/server/snapshot.cc

```diff
@@ -78,9 +78,10 @@ void SliceSnapshot::Cancel() {
   VLOG(1) << "SliceSnapshot::Cancel";
 
   CloseRecordChannel();
-  if (journal_cb_id_) {
-    db_slice_->shard_owner()->journal()->UnregisterOnChange(journal_cb_id_);
-    journal_cb_id_ = 0;
+  // Set journal_cb_id_ to 0 and unregister callback, but only once.
+  uint32_t cb_id = journal_cb_id_.load();
+  if (cb_id && journal_cb_id_.compare_exchange_strong(cb_id, 0)) {
+    db_slice_->shard_owner()->journal()->UnregisterOnChange(cb_id);
```

Comment:
I always thought the cb-id is accessed only from one thread (but possibly multiple fibers). In what case is it accessed by multiple threads?

Reply:
You're right, I overreacted with the atomics. The bug was that UnregisterOnChange(cb_id) acquires a global lock so it preempts the fiber but a simple assignment before call should be enough.